### PR TITLE
Adds support for including signing time in signedAttrs for CMS signing.

### PR DESCRIPTION
**Pull request by R4N**
Motivation:

Many CMS signing validations require the inclusion of signing time in signedAttrs (for example Apple Wallet's Pass CMS signing validation)

Modifications:

* Adds optional includeSigningTime flag to CMS.sign method (defaults to false)
* When includeSigningTime flag is true: includes signing-time (UTCTime ASN1Type), content-type, and message-digest signedAttrs and signs those instead of content bytes (as specified in RFC 5652 section 5.4)
* Adds contentType ASN1ObjectIdentifier
* Adds CMSSigningError (noTimeZone, invalidSigningDateComponent) for when includeSigningTime is true and unable to get UTC time zone or date components
* Adds tests for valid and invalid CMS signature when including signedAttrs

Result:

Providing the includeSigningTime = true flag to CMS.sign will cause signing-time, content-type, and message-digest signedAttrs to be included in the CMS signature and the signature bytes will represent signing those signedAttrs. Using CMS.sign without the flag (defaulting to false) should leave the previous CMS.sign behavior unchanged.

Resolves: https://github.com/apple/swift-certificates/issues/147

Submitted on behalf of Zetetic, LLC

### CODE REVIEW

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 1]
```diff
@@ -23,15 +28,82 @@ public enum CMS {
...
+                Calendar.Component.hour,
+                Calendar.Component.minute,
+                Calendar.Component.second,
+            ]
+            guard let timeZone = TimeZone(identifier: "UTC") else {
+                throw CMSSigningError.noTimeZone
+            }
+            var calendar = Calendar(identifier: .gregorian)
+            calendar.timeZone = timeZone
+            let nowComponents = calendar.dateComponents(desiredComponents, from: Date())
```
Comment (Lukasa) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:69]: We should avoid this. We have a `Date.utcTime` transformation that already exists.

**Review by Lukasa — CHANGES_REQUESTED** [review]
Thanks for this! I've left a few notes in the diff.

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 2]
```diff
@@ -23,15 +28,82 @@ public enum CMS {
...
+            // 1. content-type
+            // 2. message-digest
+
+            // add content-type signedAttr cms data
+            let contentTypeVal = try ASN1Any(erasing: ASN1ObjectIdentifier.cmsData)
+            let contentTypeAttribute = CMSAttribute(attrType: .contentType, attrValues: [contentTypeVal])
+            signedAttrs.append(contentTypeAttribute)
+
+            // add message-digest sha256 of provided content bytes
+            try SHA256.hash(data: bytes).withUnsafeBytes { bufferPointer in
```
[the remark this answers is not in the document]
Comment (dnadoba) [reply Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:48]: We should not need any unsafe Swift here. The returned digest conforms to `Sequence<UInt8>`. Just convert the returned bytes to an `ArraySlice`.

**Review by dnadoba — CHANGES_REQUESTED** [review]

File: Tests/X509Tests/CMSTests.swift [hunk 3]
```diff
@@ -1014,4 +1055,76 @@ extension CMS {
...
+        // 1. content-type
+        // 2. message-digest
+
+        // add content-type signedAttr cms data
+        let contentTypeVal = try ASN1Any(erasing: ASN1ObjectIdentifier.cmsData)
+        let contentTypeAttribute = CMSAttribute(attrType: .contentType, attrValues: [contentTypeVal])
+        signedAttrs.append(contentTypeAttribute)
+
+        // add message-digest sha256 of provided content bytes
+        try SHA256.hash(data: bytes).withUnsafeBytes { bufferPointer in
```
Comment (dnadoba) [inline Tests/X509Tests/CMSTests.swift:1083]: Also, here. No need for unsafe Swift.

**Comment by R4N** [comment]
Thanks for the review/feedback. Definitely good suggestions. Using Date.utcDate to construct the UTCTime from date components allows me to remove the previously defined `CMSSigningError`s as both of the defined errors were related to DateComponents construction.

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 4]
```diff
@@ -23,15 +23,77 @@ public enum CMS {
         signatureAlgorithm: Certificate.SignatureAlgorithm,
         additionalIntermediateCertificates: [Certificate] = [],
         certificate: Certificate,
-        privateKey: Certificate.PrivateKey
+        privateKey: Certificate.PrivateKey,
+        signingTime: Date? = nil
     ) throws -> [UInt8] {
+        if let signingTime = signingTime {
+            return try signWithSigningTime(
```
Comment (Lukasa) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:30]: Nit: we prefer to use explicit `self` in all places.

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 5]
```diff
@@ -23,15 +23,77 @@ public enum CMS {
...
         let signature = try privateKey.sign(bytes: bytes, signatureAlgorithm: signatureAlgorithm)
-        return try sign(
+        let signedData = try self.generateSignedData(
             signatureBytes: ASN1OctetString(signature),
             signatureAlgorithm: signatureAlgorithm,
             additionalIntermediateCertificates: additionalIntermediateCertificates,
             certificate: certificate
         )
+
+        return try serializeSignedData(signedData)
```
Comment (Lukasa) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:48]: Same nit here, prefer `self.`

File: Sources/X509/CryptographicMessageSyntax/CMSOperations.swift [hunk 6]
```diff
@@ -23,15 +23,77 @@ public enum CMS {
...
+        let signedAttrBytes = coder.serializedBytes[...]
+        let signature = try privateKey.sign(bytes: signedAttrBytes, signatureAlgorithm: signatureAlgorithm)
+        let signedData = try self.generateSignedData(
+            signatureBytes: ASN1OctetString(signature),
+            signatureAlgorithm: signatureAlgorithm,
+            additionalIntermediateCertificates: additionalIntermediateCertificates,
+            certificate: certificate,
+            signedAttrs: signedAttrs
+        )
+        return try serializeSignedData(signedData)
```
Comment (Lukasa) [inline Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:96]: Same note on `self` here.

**Comment by R4N** [comment]
Added in the explicit `self`'s for the internal methods calls within CMSOperations.

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Comment by Lukasa** [comment]
These failures aren't your fault, our allocation baselines have regressed. I'll update them.

**Comment by Lukasa** [comment]
Feel free to rebase on to the new `main` branch.

**Comment by R4N** [comment]
I believe that should fix it up with the rebase now.

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Comment by R4N** [comment]
Just rebased off of main for the latest commit that came in. The last `swift-server-bot` check had a failure because of the `CMS.sign` API method signature changing. I believe this is a false positive because of the optional `signingTime` [parameter addition](https://github.com/R4N/swift-certificates/blob/signing-time/Sources/X509/CryptographicMessageSyntax/CMSOperations.swift#L27). Let me know if anything else needs to be done on my end.

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Review by dnadoba — APPROVED** [review]
LGTM. Thanks!

**Comment by Lukasa** [comment]
Merging over the API breakage.

**Comment by dnadoba** [comment]
Note that this is in theory an API breaking change as unapplied methods do not take default values into account:
```swift
// this fails to compile now but previously compiled successfully
let ref = CMS.sign(_:signatureAlgorithm:additionalIntermediateCertificates:certificate:privateKey:)
```
We can partially workaround this limitation by keeping the old signature and introducing an overload with the new parameter. The func with the old signature would just call into the new method. 
However, this only partially solves the source break as
```swift
// error: Ambiguous use of 'CMS.sign'
let ref = CMS.sign
```
now becomes ambiguous and fails to compile as well.

I personally think this is fine though. 
@Lukasa wdyt?

**Comment by Lukasa** [comment]
Yeah I don't mind at all. These are also CMS functions so they're not guaranteed to be API stable, so this isn't very egregious.

**Pull request merged**